src/BaseKrylov.f90

```diff
@@ -821,27 +828,157 @@ subroutine qr_factorization(Q, R, info, verbosity, tol)
-         end do
-         beta = Q(j)%norm(); call Q(j)%scal(1.0_wp/beta)
-         if (abs(beta) < tolerance) then
-            if (verbose) then
-               write (*, *) "INFO : Colinear columns detected."
-               write (*, *) "       (j, beta) = (", j, ",", beta, ")."
-            end if
-         end if
-      end do
       if (verbose) then
          write (*, *) "INFO : Exiting the QR factorization with exit code info = ", info, "."
       end if
 
       return
    end subroutine qr_factorization
 
+   subroutine block_arnoldi_factorization(A, X, H, info, kstart, kend, verbosity, tol, transpose, block_size)
+
+      ! --> Optional arguments (mainly for GMRES)
+      integer, optional, intent(in) :: kstart, kend
+      logical, optional, intent(in) :: verbosity, transpose
+      real(kind=wp), optional, intent(in) :: tol
+  
+      integer :: k_start, k_end, p
+      logical :: verbose, trans
+      real(kind=wp) :: tolerance
+
+      ! --> Optional: size of blocks, default = 1
+      integer, optional, intent(in) :: block_size
+  
+      ! --> Linear Operator to be factorized.
+      class(abstract_linop), intent(in) :: A
+      ! --> Krylov basis.
+      class(abstract_vector), intent(inout) :: X(:)
+      ! --> Upper Hessenberg matrix.
+      real(kind=wp), intent(inout) :: H(:, :)
+      ! --> Information.
+      integer, intent(out) :: info ! info < 0 : The k-step Arnoldi factorization failed.
+      ! info = 0 : The k-step Arnoldi factorization succeeded.
+      ! info > 0 : An invariant subspace has been computed after k=info steps.
+      ! --> Miscellaneous
+      real(kind=wp) :: beta
+      real(kind=wp), allocatable :: res(:)
+      integer :: k, i, kdim
+      integer :: kpm, kp, kpp
+      
+      ! --> Deals with optional non-unity block size
+      p   = optval(block_size, 1)
+      allocate(res(1:p))
+  
+      info = 0
+
+      ! --> Check dimensions.
+      kdim = (size(X) - p)/p ; call assert_shape(H, [p*(kdim+1), p*kdim], "arnoldi_factorization", "H")
+  
+      ! --> Deals with the optional arguments.
+      k_start   = optval(kstart, 1)
+      k_end     = optval(kend, kdim)
+      verbose   = optval(verbosity, .false.)
+      tolerance = optval(tol, atol)
+      trans     = optval(transpose, .false.)
+  
+      ! --> Arnoldi factorization.
+      block_arnoldi: do k = k_start, k_end
+         ! --> Counters
+         kpm = (k-1)*p
+         kp  = kpm + p
+         kpp = kp  + p
+         ! --> Matrix-vector products.
+         if (trans) then
+            do i = 1,p
+               call A%rmatvec(X(kpm+i), X(kp+i))
+            enddo
+         else
+            do i = 1,p
+               call A%matvec(X(kpm+i), X(kp+i))
+            enddo
+         endif
+         ! --> Update Hessenberg matrix w.r.t. existing vectors
+         call block_update_hessenberg_matrix(H, X, k, p)
+         ! --> Orthogonalize current vectors
+         call qr_factorization(X(kp+1:kpp),H(kp+1:kpp,kpm+1:kp),info)
```

Comment:
Why do you need to call `qr_factorization` here ?
Isn't the orthogonalization already taken care of when updating the block Hessenberg matrix ?

Reply:
It comes down to style and I took the same structure as in your arnoldi.

update hessenberg actually just fills in the rightmost column(s) of the hessenberg matrix up to and including the diagonal, i.e. orthogonalising the new vector(s) against all the previous ones. 

After this stage, the normal arnoldi would proceed to normalise the new vector and place the norm on the subdiagonal of the hessenberg matrix.
In the block arnoldi case, this normalisation is actually a QR of the p new columns.

In summary, the following parts of the code are conceptually identical.

`arnoldi_factorization`:
```
call update_hessenberg_matrix(H, X, k)
beta = X(k + 1)%norm(); H(k + 1, k) = beta
[...]
call X(k + 1)%scal(1.0_wp/beta)
```
`block_arnoldi_factorization`:
```
call block_update_hessenberg_matrix(H, X, k, p)
call qr_factorization(X(kp+1:kpp),H(kp+1:kpp,kpm+1:kp),info)
```

I understand that you separated the two steps in `arnoldi_factorization` in order to be able to do the convergence test for beta separately. You could do this in `block_arnoldi_factorization` too by creating a temporary R matrix in the QR, find the smallest beta and then add it to the Hessenberg matrix.

Should we do it this way?

Reply:
Ah no ok, it's fine. I did not realized it boiled down to the same thing.